src/http-gateways/trustless-gateway.md

```diff
@@ -301,6 +302,31 @@ of their presence in the DAG or the value assigned to the "dups" parameter, as
 the raw data is already present in the parent block that links to the identity
 CID.
 
+## CAR `meta` (content type parameter)
+
+The `meta` parameter allows clients to request the server to include additional metadata about the
+CAR to be included at the end of the response body.
+
+This parameter can be used with `version=1` only.
+
+When the parameter is not set, the server must not add any extra CAR blocks to the response.
+
+The metadata block is a regular CAR block with the following properties:
+
+- CID specifies multicodec `car-metada` (0x04ff), see
+  [multicodec#334](https://github.com/multiformats/multicodec/pull/334).
+
+- The payload contains metadata encoded as DAG-CBOR.
+
+The metadata MUST include the following fields:
+
+- `len` - byte length of the CAR data (excluding the metadata block)
+- `b3h` - Blake3 hash (checksum) of the CAR data (excluding the metadata block).
+- `b3h_sig` - A signature over `<len><b3h><request>` using server's Ed2559 identity.
+  - `len` is encoded as `varint`,
+  - `b3h` is encoded as 32 bytes,
+  - The effective query as executed by the gateway. This query is the request url - path and query string arguments.
```

Comment:
@lidel what is the reasoning behind splitting up `dag_params` and `car_params`? Could we instead go for `content_path` and `query_params` to keep it simple and generic (allowing for other query params)?

Reply:
I believe that was a conscious design choice, to avoid mixing data selector with details of the transport format (not everything is in URL query params):

- `dag_params` are about what user data was selected, not tied to any specific transport (could be applied to something other than CAR)
   - these are things that land in URL query
- `car_params` are specific to CAR container format, they do not change the user data that was selected, only the way it is represented when sent as CAR  
   -  these are things that land in Accept/Content-Type headers

@patrickwoodhead  that being said, if you want to simplify, this IPIP could go with a single dag-json map named `response_params` (to avoid confusion with URL query params, and account for the fact that server may ignore some of request params when producing a response)

Reply:
How about using `retrieval_params` instead of `response_params`? In my mind, response parameters don't describe "what user data was selected".

What is our motivation in SPARK: 
We want the gateway to describe what exactly the client is retrieving (CID, subpath, dag params, car params) and provide a signature over that.

If two SPARK checker clients submit a metadata block with the same retrieval parameters (CID, subpath, dag params, car params) then we want:
- To be able to verify that the clients retrieved the DAG (sub)tree they were expected to retrieve.
- Confidence that both clients made the same request to the gateway (semantically?) and were supposed to receive exactly the same response from the gateway.

> `content_path` - requested content path

When the client requests `GET /ipfs/bafy1234/cat.jpg`, what is `content_path`?
- `/ipfs/bafy1234/cat.jpg`
- `bafy1234/cat.jpg`
- `/cat.jpg`
- something else?

We need the metadata block to describe both the CID requested (`bafy1234`) and the resource subpath if specified (`/cat.jpg`).